src/spyglass/spikesorting/v1/recompute.py

```diff
@@ -0,0 +1,527 @@
+        _ = this_env.pop("spyglass", None)  # ignore spyglass version
+        return this_env
+
+    def namespace_dict(self, type_map: TypeMap):
+        """Remap namespace names to hyphenated field names for DJ compatibility."""
+        name_cat = type_map.namespace_catalog
+        return {
+            field: name_cat.get_namespace(field).get("version", None)
+            for field in name_cat.namespaces
+        }
+
+    def make(self, key):
+        """Inventory the namespaces present in an analysis file."""
+        query = SpikeSortingRecording() & key
+        if not len(query) == 1:
+            raise ValueError(
+                f"SpikeSortingRecording & {key} has {len(query)} "
+                + f"matching entries: {query}"
+            )
+
+        parent = query.fetch1()
+        path = AnalysisNwbfile().get_abs_path(parent["analysis_file_name"])
+
+        nwb_deps = get_file_namespaces(path).copy()
+
+        with h5py_File(path, "r") as f:
+            script = f.get("general/source_script")
+            if script is not None:  # after `=`, remove quotes
+                script = str(script[()]).split("=")[1].strip().replace("'", "")
+            nwb_deps["spyglass"] = script
+
+        self.insert1(dict(key, nwb_deps=nwb_deps), allow_direct_insert=True)
+
+
+# Initializing table to prevent recompute of cached properties
+REC_VER_TBL = RecordingRecomputeVersions()
+
+
+@schema
+class RecordingRecomputeSelection(SpyglassMixin, dj.Manual):
+    definition = """
+    -> RecordingRecomputeVersions
+    -> UserEnvironment
+    rounding=4: int # rounding for float ElectricalSeries
+    ---
+    logged_at_creation=0: bool # whether the attempt was logged at creation
+    """
+
+    # --- Insert helpers ---
+    @cached_property
+    def default_rounding(self) -> int:
+        """Return the default rounding for ElectricalSeries data."""
+        return int(self.heading.attributes["rounding"].default)
+
+    @cached_property
+    def env_dict(self):
+        return USER_TBL.insert_current_env()
+
+    def insert(self, rows, limit=None, at_creation=False, **kwargs) -> None:
+        """Custom insert to ensure dependencies are added to each row."""
+        if not rows:
+            logger.info("No rows to insert.")
+            return
+        if not isinstance(rows, (list, tuple)):
+            rows = [rows]
+        if not isinstance(rows[0], dict):
+            raise ValueError("Rows must be a list of dicts")
+
+        inserts = []
+        for row in rows:
+            row_pk = self.dict_to_pk(row)
+            if not REC_VER_TBL._has_matching_env(row_pk):
+                continue
+            full_key = self.dict_to_full_key(row)
+            full_key.update(dict(self.env_dict, logged_at_creation=at_creation))
+            inserts.append(full_key)
+
+        super().insert(inserts, **kwargs)
+
+    def attempt_all(
+        self,
+        restr: dict = True,
```

Comment:
type should be `Union[bool, dict]`

Reply:
I'm thinking `Optional[dict]` is more accurate. I wouldn't want someone setting `restr=False`

Reply:
So optional meaning True or None or dict? Can you catch `restr=False` then just in case?

Reply:
I'll add a warning if the restriction doesn't yield inserts